### ISSUE TITLE
analyze and prepare concept how to fix Overflow numbers

### ISSUE THREAD

**Issue by koonopek** [redstone-finance/redstone-oracles-monorepo#491]
Two potential solutions:

# Using the library for decimals
(+) allows for the arbitrary size of numbers (we can support a wider range of dataFeed (is there any?))
(-) potential problems with serialization/deserializations

## BN.js
- most popular
- ethers uses this under the hood
- doesn't support decimals

## decimal.js
- https://github.com/MikeMcl/decimal.js/
- slow

## decimal.js-light
- https://www.npmjs.com/package/decimal.js-light
- last published 2 years ago 
- slow
- better then decimal.js cause smaller and doesn't support Infinity

## exact-math
- https://www.npmjs.com/package/exact-math#object-config-properties
- doesn't support comparing operands

# Using native numbers in safe ranges 
Almost ready implementation showing the idea: https://github.com/redstone-finance/redstone-oracles-monorepo/pull/495
If we choose this option I would change the implementation to encapsulate it in a class like SafeNumber, because now it has very pure DX (dev has to call `clamp(assertSafeNumber())` on results operating on numbers  ...).

(+) no extra dependency
(+) allocate 8-10x less memory (depends on lib) - https://gist.github.com/koonopek/887be8f20852b5766331301065966e98 (200_000 numbers takes 50 MB of RAM)
```bash
One iteration allocates 2000 values (simulation of oracle-node work)
┌─────────┬────────────┬────────────┐
│ (index) │ iterations │  MBUsage   │
├─────────┼────────────┼────────────┤
│    0    │     1      │  '0.7971'  │
│    1    │     10     │  '7.8537'  │
│    2    │    100     │ '52.7445'  │
│    3    │    1000    │ '512.6969' │
└─────────┴────────────┴────────────┘
For native numbers:
┌─────────┬────────────┬───────────┐
│ (index) │ iterations │  MBUsage  │
├─────────┼────────────┼───────────┤
│    0    │     1      │ '0.0968'  │
│    1    │     10     │ '0.8782'  │
│    2    │    100     │ '6.1866'  │
│    3    │    1000    │ '58.3203'
```
(+) BN.js 20 times slower, decimal.js-light >200 times slower than the native numbers. 
```bash
simulation consists of 10 probes for 100_000 iteration in loop of given operation described in "what" column
All numbers are in [MS]
┌─────────┬─────────────────────────────────┬────────────┬──────────┬───────────┬───────────┬─────────────────────┐
│ (index) │             what:               │   avg:     │  dev:    │   min:    │   max:    │    memory [MB]:     │
├─────────┼─────────────────────────────────┼────────────┼──────────┼───────────┼───────────┼─────────────────────┤
│    0    │      'BN inplace add int'       │  '4.5124'  │ '1.8663' │ 3.386417  │ 9.349084  │ 0.6143646240234375  │
│    1    │          'BN add int'           │  '9.3104'  │ '2.5808' │ 8.034792  │ 16.14375  │  1.53643798828125   │
│    2    │             '+ int'             │  '1.1336'  │ '1.4071' │ 0.431708  │ 4.272875  │ 0.2372894287109375  │
│    3    │      'BN inplace mul int'       │  '7.7740'  │ '2.7224' │   6.801   │ 15.938834 │ 0.6827621459960938  │
│    4    │          'BN mul int'           │  '8.4868'  │ '2.4947' │ 7.606917  │ 15.970167 │ 1.3961563110351562  │
│    5    │             '* int'             │  '0.8100'  │ '1.1268' │ 0.432708  │ 4.190333  │ 0.20915985107421875 │
│    6    │     'BN inplace div float'      │ '14.9799'  │ '2.8736' │ 13.942917 │  23.5995  │ 0.5473403930664062  │
│    7    │           'BN div  '            │  '6.8933'  │ '1.4211' │ 6.343291  │ 11.154083 │ 1.1937713623046875  │
│    8    │            '/ float'            │  '0.9288'  │ '1.4834' │ 0.432375  │ 5.378917  │ 0.36865997314453125 │
│    9    │     'BN inplace add float'      │  '3.4651'  │ '1.0332' │  3.08675  │ 6.564209  │   0.8966064453125   │
│   10    │           'BN add  '            │  '5.5668'  │ '3.2795' │ 4.447666  │ 15.405209 │  1.430572509765625  │
│   11    │            '+ float'            │  '0.4951'  │ '0.1614' │ 0.439375  │ 0.979125  │ 0.39942169189453125 │
│   12    │        'Decimal add int'        │ '36.2406'  │ '1.6759' │ 35.366166 │  41.0935  │  1.06817626953125   │
│   13    │        'Decimal mul int'        │ '55.3209'  │ '1.4566' │ 54.669625 │ 59.674208 │ 1.5849761962890625  │
│   14    │         'Decimal div  '         │ '134.1205' │ '1.4159' │  133.489  │  138.356  │ 0.23082733154296875 │
│   15    │         'Decimal add  '         │ '96.5207'  │ '0.8796' │ 95.85325  │ 99.030208 │ 0.35926055908203125 │
│   16    │ 'Pre allocated Decimal add int' │  '8.9724'  │ '0.1446' │ 8.884458  │   9.401   │ 1.1334991455078125  │
│   17    │ 'Pre allocated Decimal mul int' │ '26.8695'  │ '0.3332' │ 26.691292 │ 27.85575  │ 0.6820907592773438  │
│   18    │  'Pre allocated Decimal div  '  │ '60.8873'  │ '0.3288' │ 60.650834 │ 61.805791 │ 0.39914703369140625 │
│   19    │  'Pre allocated Decimal add  '  │ '24.6667'  │ '0.1721' │ 24.558375 │ 25.132625 │   1.5394287109375   │
└─────────┴─────────────────────────────────┴────────────┴──────────┴───────────┴───────────┴─────────────────────┘
```
- https://github.com/redstone-finance/redstone-oracles-monorepo/issues/486

(+) much easier to integrate into the current code space
(-) own implementation (security concerns)
(-) small bounded range of supported numbers <0, 2^53 - 1>

 TLDR:
 Trade-off is the performance for security wider number range and convenience.

- [ ] Prepare documentation on numbers (especially their constraints) in Redstone protocol:

**Comment by koonopek** [comment]
@hatskier @cehali
Any thoughts?

**Comment by cehali** [comment]
This could be a good trade-off, implementation is almost ready. Also, I wonder if we should not accept numbers bigger than JS max from the fetchers and then we will notice that a particular token is not delivered.

**Comment by koonopek** [comment]
@cehali  Yeah, good idea, validation is happening after fetchers
